src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+
+/-!
+# Circulant matrices
+
+This file contains the definition and basic results about circulant matrices.
+Given a vector `v : I → α` indexed by a type that is endowed with subtraction,
+`circulant v` is the matrix whose `(i,j)`th entry is `v (i = j)`.
+
+## Main results
+
+- `matrix.circulant`: the circulant matrix generated by a given vector `v : I → α`.
```

Comment:
I would say that `circulant_mul_comm` and `circulant_mul` are worthy of being listed as main results too (unlike most of the others which are trivially true). Don't forget the `matrix.` prefix!